File: src/org/jabby/blazenav/NavActionBase.java
package org.jabby.blazenav;

import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.LangDataKeys;
import com.intellij.openapi.actionSystem.PlatformDataKeys;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.FoldRegion;
import com.intellij.openapi.editor.markup.HighlighterLayer;
import com.intellij.openapi.editor.markup.HighlighterTargetArea;
import com.intellij.openapi.editor.markup.RangeHighlighter;
import com.intellij.openapi.editor.markup.TextAttributes;
import com.intellij.psi.PsiFile;
import com.intellij.ui.JBColor;
import com.intellij.util.Range;
import org.apache.commons.lang.StringUtils;

import java.awt.*;
import java.util.ArrayList;
import java.util.List;

abstract class NavActionBase extends ActionBase {
    @Override public final void update(AnActionEvent e) {
        PsiFile psiFile = e.getData(LangDataKeys.PSI_FILE);
        Editor editor = e.getData(PlatformDataKeys.EDITOR);
        e.getPresentation().setEnabled(psiFile != null && editor != null);
    }

    @Override public final void actionPerformed(AnActionEvent e) {
        PsiFile psiFile = e.getData(LangDataKeys.PSI_FILE);
        Editor editor = e.getData(PlatformDataKeys.EDITOR);

        if (psiFile != null && editor != null) actionPerformed(psiFile, editor);
    }

    abstract void actionPerformed(PsiFile psiFile, Editor editor);

    Runnable unfold(final Editor editor) {
        FoldRegion[] folds = editor.getFoldingModel().getAllFoldRegions();
        final List<FoldRegion> collapsed = new ArrayList<FoldRegion>(folds.length);
        for (FoldRegion fold : folds) { if (!fold.isExpanded()) collapsed.add(fold); }

        editor.getFoldingModel().runBatchFoldingOperation(new Runnable() {
            @Override public void run() {
                for (FoldRegion fold : collapsed) fold.setExpanded(true);
            }
        });

        return new RunnableOnce() {
            @Override protected void runOnce() {
                editor.getFoldingModel().runBatchFoldingOperation(new Runnable() {
                    @Override public void run() {
                        for (FoldRegion fold : collapsed) fold.setExpanded(false);
                    }
                });
            }
        };
    }

    Runnable disableHighlighting(final Editor editor, int textLength) {
        TextAttributes ta = new TextAttributes();
        ta.setForegroundColor(JBColor.GRAY);
        ta.setEffectColor(JBColor.GRAY);
        ta.setErrorStripeColor(null);
        ta.setEffectType(null);

        final RangeHighlighter highlighter = highlight(editor, 0, textLength, ta);

        return new RunnableOnce() {
            @Override protected void runOnce() { editor.getMarkupModel().removeHighlighter(highlighter); }
        };
    }

    private RangeHighlighter highlight(Editor editor, int startOffset, int endOffset, TextAttributes ta) {
        return editor.getMarkupModel()
                .addRangeHighlighter(
                        startOffset, endOffset, HighlighterLayer.SELECTION, ta, HighlighterTargetArea.EXACT_RANGE);
    }

    Range<Integer> getVisibleLinesRange(Editor editor) {
        Rectangle visibleArea = editor.getScrollingModel().getVisibleArea();
        int lineHeight = editor.getLineHeight();

        int fromLine = getLineNo(visibleArea.y, lineHeight) + 1;
        int toLine = getLineNo(visibleArea.y + visibleArea.height, lineHeight) - 1;

        return new Range<Integer>(fromLine, toLine);
    }

    private int getLineNo(int topOffset, int lineHeight) {
        int ratio = topOffset / lineHeight;
        return topOffset % lineHeight == 0 ? ratio : ratio + 1;
    }

    int getFromOffset(String text, int fromLine) {
        return fromLine > 1 ? StringUtils.ordinalIndexOf(text, "\n", fromLine - 1) + 1 : 0;
    }

    int getToOffset(String text, int toLine) {
        int toOffset = StringUtils.ordinalIndexOf(text, "\n", toLine);
        if (toOffset == -1) toOffset = text.length();
        return toOffset;
    }
}

File: src/org/jabby/blazenav/NavActionBase.kt
package org.jabby.blazenav

import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.markup.TextAttributes
import com.intellij.ui.JBColor
import java.awt.Font
import com.intellij.openapi.editor.markup.HighlighterLayer
import com.intellij.openapi.editor.markup.HighlighterTargetArea
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.PlatformDataKeys
import org.apache.commons.lang.StringUtils

abstract class NavActionBase: ActionBase() {
    data class Bounds(val fromLine: Int, val toLine: Int)

    protected abstract fun actionPerformed(editor: Editor)

    public override final fun update(e: AnActionEvent?) {
        val editor = e!!.getData(PlatformDataKeys.EDITOR)
        e.getPresentation().setEnabled(editor != null)
    }

    public override final fun actionPerformed(e: AnActionEvent?) {
        val editor = e!!.getData(PlatformDataKeys.EDITOR)
        if (editor != null) actionPerformed(editor)
    }

    protected fun fade(editor: Editor, textLength: Int): (() -> Unit) {
        val ta = TextAttributes(JBColor.GRAY, null, JBColor.GRAY, null, Font.PLAIN);
        ta.setErrorStripeColor(null);

        val fader = editor.getMarkupModel().addRangeHighlighter(
                0, textLength, HighlighterLayer.SELECTION, ta, HighlighterTargetArea.EXACT_RANGE)

        return {() -> editor.getMarkupModel().removeHighlighter(fader) }
    }

    protected fun unfold(editor: Editor): (() -> Unit) {
        val folds = editor.getFoldingModel().getAllFoldRegions()
        val collapsed = folds.filter { !it.isExpanded() }

        editor.getFoldingModel().runBatchFoldingOperation { collapsed.forEach { it.setExpanded(true) } }

        return {
            () ->
            editor.getFoldingModel().runBatchFoldingOperation { collapsed.forEach { it.setExpanded(false) } }
        }
    }

    protected fun getVisibleBounds(editor: Editor): Bounds {
        val area = editor.getScrollingModel().getVisibleArea()
        val lineHeight = editor.getLineHeight()

        fun getLineNo(topOffset: Int) = Math.ceil((area.y + topOffset) / lineHeight.toDouble()).toInt()

        return Bounds(getLineNo(0) + 1, getLineNo(area.height) - 1)
    }

    protected fun getFromOffset(text: String, fromLine: Int): Int =
            if (fromLine > 1) StringUtils.ordinalIndexOf(text, "\n", fromLine - 1) + 1 else 0

    protected fun getToOffset(text: String, toLine: Int): Int {
        val toOffset = StringUtils.ordinalIndexOf(text, "\n", toLine)
        return if (toOffset == -1) text.length() else toOffset
    }
}
